src/hotspot/share/opto/subnode.cpp

```diff
@@ -1899,3 +1899,50 @@ const Type* SqrtFNode::Value(PhaseGVN* phase) const {
   if( f < 0.0f ) return Type::FLOAT;
   return TypeF::make( (float)sqrt( (double)f ) );
 }
+
+static jlong reverse_bits(jlong val) {
+  jlong res = ((val & 0xF0F0F0F0F0F0F0F0L) >> 4) | ((val & 0x0F0F0F0F0F0F0F0F) << 4);
+  res = ((res & 0xCCCCCCCCCCCCCCCCL) >> 2) | ((res & 0x3333333333333333L) << 2);
+  res = ((res & 0xAAAAAAAAAAAAAAAAL) >> 1) | ((res & 0x5555555555555555L) << 1);
+  return res;
+}
+
+const Type* ReverseINode::Value(PhaseGVN* phase) const {
+  const Type *t1 = phase->type( in(1) );
+  if (t1 == Type::TOP) {
+    return Type::TOP;
+  }
+  const TypeInt* t1int = t1->isa_int();
+  if (t1int && t1int->is_con()) {
+    jint res = reverse_bits(t1int->get_con());
```

Comment:
This is clearly wrong, if I understand correctly that ReverseI is an intrinsic for `Integer::reverse`, which reverses not only bits in bytes but also the bytes themselves.

Also, it seems noxious (IIUC) because it widens t1int's con to 64 bits with sign extension, does some sort of 64-bit reversal operation, and then chops off the high order bits assigning to `res`.

The root cause here is that `reverse_bits` should be properly factored out and maintained in its own modular header file, akin to `utilities/powerOfTwo.hpp`.  I suggest `utilities/moveBits.hpp`, to hold bits-in-bytes reversal, byte reversal, full reversal, compress/expand (with which it should be grouped) and who knows what other future permutations.

I say it's a root cause in that something randomly coded in the bowels of `subnode.cpp` cannot be properly reviewed in isolation, especially if it is misnamed.

Also, the nearly identical use of the HD algorithm in `code/compressedStream.cpp` should use the new version in the header file.

The new hand-written assembly code for this changeset should refer the reader to utilities/moveBits.hpp as well.